src/main/java/com/google/devtools/build/lib/analysis/constraints/TopLevelConstraintSemantics.java

```diff
@@ -136,6 +155,83 @@ private MissingEnvironment(Label environment, RemovedEnvironmentCulprit culprit)
         continue; // This target doesn't participate in constraints.
       }
 
+      Object target_compatible_with_object =
+          target.getAssociatedRule().getAttributeContainer().getAttr("target_compatible_with");
```

Comment:
Let's get this through NonconfigurableAttributeMapper instead (which is the recommended way to get attribute values and applies proper typing).

You may have to cast the Target to Rule to do that.

Reply:
I'll give that a try, thanks!